# Lab 1: Basic Python Programming

Yuil Tripathee `65070503480`

## 1. Basic usage

> John Doe is a 29 years-old system engineer who earns ฿41500.00 a month.

Create and assign variables to store this person's information (name, age, position and salary).

In [20]:
# Write your code here
name = "John Doe"
age = 29
position = "System Engineer"
salary = 41500.00

What is the type of each variables?

In [15]:
# Write your code here
print(type(name))
print(type(age))
print(type(position))
print(type(salary))

<class 'str'>
<class 'int'>
<class 'str'>
<class 'float'>


The manager decides to give John a 7% raise. Update his salary.

In [23]:
# Write your code here
raise_p = 7

raise_v = salary * (raise_p / 100);

salary_latest = salary + raise_v;

print(salary_latest)

44405.0


Prints his information again with his new salary.

In [24]:
# Write your code here
print(salary_latest)

44405.0


Now, he decides to resign. Delete his information from the system.

In [28]:
# Write your code here
del name
del age
del position
del salary

NameError: name 'name' is not defined

## 2. Variable and Expression

**2.1** Write a code to convert temperature unit from celcius to other units

In [30]:
C = 34.5

**Fahrenheit**  
#### $\frac{C}{5} = \frac{F-32}{9}$

In [33]:
F = ((9*C)/5) + 32
print(F)

94.1


**Kelvin**  
$K = C + 273.15$

In [37]:
K = C + 273.15
print(K)

307.65


**Rømer**  
#### $Ro = \frac{C \times 21}{40} + 7.5$

In [40]:
Ro = ((C * 21)/40) + 7.5;
print(Ro)

25.6125


## 3. Multi-item variables
#### List

In [41]:
names = ['Thomas', 'Kate', 'Mike', 'Amelia', 'James', 'Megan']

Create new variable call `new_name` which takes input name of the user.

In [43]:
new_name = input('Enter your name: ')

Enter your name:  Jerry


Insert `new_name` into `names` list.

In [44]:
# Write your code here
names.append(new_name)

Select your name from the list

In [47]:
# Write your code here
print(names[-1])

Jerry


Merge `another_names` into `names`.

In [49]:
another_names = ['Peter', 'Steve', 'Sam', 'Charlotte']

In [50]:
# Write your code here
names.extend(another_names)
print(names)

['Thomas', 'Kate', 'Mike', 'Amelia', 'James', 'Megan', 'Jerry', 'Peter', 'Steve', 'Sam', 'Charlotte']


Change `Amelia`'s name to `Amy`

In [53]:
# Write your code here
if 'Amelia' in names:
    i_amelia = names.index('Amelia')
    names[i_amelia] = 'Amy'

print(names)

['Thomas', 'Kate', 'Mike', 'Amy', 'James', 'Megan', 'Jerry', 'Peter', 'Steve', 'Sam', 'Charlotte']


#### Dictionary

In [54]:
capital_city = {'England':'London',
                'Spain':'Madrid',
                'Japan':'Tokyo',
                'Australia':'Sydney',
                'Germany':'Berlin',
               }

Add a record `Thailand` and it's capital city to this dictionary

In [55]:
# Write your code here
capital_city['Thailand'] = 'Bangkok'

You may notice that the capital city of `Australia` is wrong. It should be `Canberra`. Correct this mistake.

In [56]:
# Write your code here
capital_city['Australia'] = 'Canberra'
print(capital_city)

{'England': 'London', 'Spain': 'Madrid', 'Japan': 'Tokyo', 'Australia': 'Canberra', 'Germany': 'Berlin', 'Thailand': 'Bangkok'}


## 4. Control Flows and conditional statements

### if...elif...else

**1.** Define a variable to get input age from user.

In [60]:
age = int(input("Enter your age: "))

Enter your age:  65


Write a series of if...elif...else statement that categorize input age into following groups:
> Babies: 0-2 years old  
Children: 3-12 years old  
Teenager: 13-19 years old  
Young Adults: 20-29 years old  
Middle-aged Adults: 30-45 years old  
Old Adult: 46-59 years old  
Elderly: Above 60 years old  

In [62]:
# Write your code here
if age <= 2:
    category = "Babies"
elif 3 <= age <= 12:
    category = "Children"
elif 13 <= age <= 19:
    category = "Teenager"
elif 20 <= age <= 29:
    category = "Young Adults"
elif 30 <= age <= 45:
    category = "Middle-aged Adults"
elif 46 <= age <= 59:
    category = "Old Adult"
else:
    category = "Elderly"

print("You are: ", category)

You are:  Elderly


### Looping

**1.** Write a code to create a multiplication table of an input number (multiplier from 1-12).

In [63]:
# Write your code here
num = int(input("Enter a number to create a multiplication table: "))

for i in range(1, 13):
    res = num * i
    print(f"{num} x {i} = {res}")


Enter a number to create a multiplication table:  12


12 x 1 = 12
12 x 2 = 24
12 x 3 = 36
12 x 4 = 48
12 x 5 = 60
12 x 6 = 72
12 x 7 = 84
12 x 8 = 96
12 x 9 = 108
12 x 10 = 120
12 x 11 = 132
12 x 12 = 144


**2.** Write a code that construct the following pattern.  

In [64]:
# Write your code here
rows = int(input("input: "))

print("output: ")
# Construct the pattern
for i in range(1, rows + 1):
    print("*" * i)

input:  5


output: 
*
**
***
****
*****


**3.** Creates a loop to print `I love <programming language>!` except for Assembly, print `Not you, Assembly`.

In [65]:
languages = ['C/C++', 'Python', 'R', 'Java', 'SQLs', 'Assembly', 'Go', 'Rust', 'Kotlin']

In [66]:
# Write your code here
for language in languages:
    if language == "Assembly":
        print("Not you, Assembly.")
    else:
        print("I love {}!".format(language))

I love C/C++!
I love Python!
I love R!
I love Java!
I love SQLs!
Not you, Assembly.
I love Go!
I love Rust!
I love Kotlin!


Student's remarks: 
But actually, I love Assembly.

**4.** Write a code to print every number from 1 to 25 except the one that is divisible by 3.

In [67]:
# Write your code here
for i in range(1, 26):
    if i % 3 != 0:
        print(i)


1
2
4
5
7
8
10
11
13
14
16
17
19
20
22
23
25


**5.** Write a code that finds the number that is divisible by 7 in a given range.

In [72]:
lower_bound = 1
upper_bound = 100
divisor = 7

result = []

In [73]:
# Write your code here
for i in range(lower_bound, upper_bound + 1):
    if i % 7 == 0:
        result.append(i)

print(result)

[7, 14, 21, 28, 35, 42, 49, 56, 63, 70, 77, 84, 91, 98]


**6.** Write a code that construct the following pattern.  

In [ ]:
input: 5  
output:  
*#####
**####
***###
****##
*****#

input: 10  
output:  
*##########  
**#########  
***########  
****#######  
*****######  
******#####  
*******####  
********###  
*********##  
**********#  

In [78]:
# Write your code here
# Get input from the user
rows = int(input("input: "))

print("output: ")
# Create the pattern
for i in range(1, rows + 1):
    # Print '*' followed by '#' based on the row number
    print('*' * i + '#' * (rows - i))


input:  10


output: 
*#########
**########
***#######
****######
*****#####
******####
*******###
********##
*********#
**********


## 5. Functions

**1.** Define a function `average` that takes arbitrary number of arguments and calculate the mean of input.

In [80]:
# Write your code here
def average(*args):
    # validation
    if not args:
        return 0  
        
    # mean
    mean = sum(args) / len(args)
    return mean

# demo
result = average(1, 2, 3, 4)
print("Mean:", result)


Mean: 2.5


**2.** Define a function `sumproduct` that takes 2 equal-sized lists and calculate sum of the products of two lists.  
It should look like this:  
> sumproduct([1,2,3],[4,5,6])  
output: 32

(1 * 4) + (2 * 5) + (3 * 6) = 32

In [81]:
# Write your code here
def sumproduct(list1, list2):
    # validation
    if len(list1) != len(list2):
        return "Error: Lists must be of equal size"

    # work
    result = sum(x * y for x, y in zip(list1, list2))
    return result

# demo
out = sumproduct([1, 2, 3], [4, 5, 6])
print("Output:", out)


Output: 32


**3.** Define a function `fibonacci` that returns Fibonacci number at `n` position.  
A Fibonacci number at position `n` is defined by `F(n) = F(n-1) + F(n-2)`. Where `F(0) = 0` and `F(1) = 1`  
**Example:** `0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, ...`  

In [86]:
# Write your code here
def F(n):
    # Base cases
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        # Recursive call
        return F(n - 1) + F(n - 2)

# Example usage
position = 10
result = F(position)
print("Fibonacci number at position", position, "is:", result)


Fibonacci number at position 10 is: 55


**4.** Define a function `is_palindrome` that takes input string and check whether it is a palindrome or not.  
A string is a palindrome if it reads the same forward and backwards.  
**Example:** `madam`, `race car`, `borrow or rob`, `amore roma`, `never odd or even`  
Do not consider whitespace. Use `str.replace(' ', '')` to remove whitespace from your string.  
Case-insensitive. You can turn everything into lower or uppercase using `str.lower()` or `str.upper()`  
**Hint:** you can reverse the string using `[::-1]` slice.

In [87]:
str1 = "radar" # palindrome
str2 = "rotator" # palindrome
str3 = "lemon" # not palindrome

In [90]:
# Write your code here
def is_palindrome(input_str):
    # cleaning: remove whitespace + all lowercase
    cleaned_str = input_str.replace(' ', '').lower()

    # forward == reverse means palindrome
    return cleaned_str == cleaned_str[::-1]

# demo
test_samples = [str1, str2, str3]

for sample in test_samples:
    result = is_palindrome(sample)
    print(f'"{sample}" is a palindrome: {result}')


"radar" is a palindrome: True
"rotator" is a palindrome: True
"lemon" is a palindrome: False


**5.** An `anagram` is a word or phrase formed by rearranging the letters of a different word or phrase.  
Define a function `is_anagram` that takes in 2 strings and check whether it is possible to compose a second string using letters in the first string or not.  
**Example:** `Tom Marrvolo Riddle` can be rearraged into `I am Lord Voldermort`  
`Meaning of Life` can be rearranged into `Engine of a Film`  
Do not consider whitespace. Use `str.replace(' ', '')` to remove whitespace from your string.  
Case-insensitive. You can turn everything into lower or uppercase using `str.lower()` or `str.upper()`  
Returns only `True` of `False`

In [101]:
# Write your code here
str1 = "Meaning of Life"
str2 = "Engine of a Film"

sample_1 = "Tom Marrvolo Riddle" # just for harry potter fun trivia
sample_2 = "I am Lord Voldermort"

def is_anagram(str1, str2):
    # cleaning: remove whitespace + all lowercase
    cleaned_str1 = str1.replace(' ', '').lower()
    cleaned_str2 = str2.replace(' ', '').lower()

    # sorted & same means anagram
    return sorted(cleaned_str1) == sorted(cleaned_str2)

test_samples = [[str1, str2], [sample_1, sample_2]]

for sample in test_samples:
    if is_anagram(sample[0], sample[1]):
        print(f'"{sample[0]}" and "{sample[1]}" are anagrams.')
    

"Meaning of Life" and "Engine of a Film" are anagrams.
"Tom Marrvolo Riddle" and "I am Lord Voldermort" are anagrams.


---